core/src/main/scala/quasar/destination/s3/S3Destination.scala

```diff
@@ -20,51 +20,80 @@ import slamdata.Predef._
 
 import quasar.api.destination.DestinationType
 import quasar.api.resource.ResourcePath
+import quasar.api.push.OffsetKey
 import quasar.blobstore.s3.Bucket
 import quasar.blobstore.paths.{BlobPath, PathElem}
-import quasar.connector.{MonadResourceErr, ResourceError}
+import quasar.connector.{AppendEvent, DataEvent, MonadResourceErr, ResourceError}
 import quasar.connector.destination.{ResultSink, UntypedDestination}
 import quasar.connector.render.RenderConfig
 import quasar.contrib.pathy.AFile
 
 import cats.data.NonEmptyList
 import cats.effect.{Concurrent, ContextShift}
 
-import cats.implicits._
+import cats.syntax.all._
 
 import fs2.{Pipe, Stream}
 
+import org.slf4s.Logger
 import pathy.Path
+import scalaz.syntax.show._
+import skolems.∀
 
-final class S3Destination[F[_]: Concurrent: ContextShift: MonadResourceErr](
-  bucket: Bucket, uploadImpl: Upload[F])
+final class S3Destination[F[_]: ContextShift: MonadResourceErr](
+  logger: Logger, bucket: Bucket, prefixPath: Option[PrefixPath], uploadImpl: Upload[F], mkPostfix: F[String])(implicit F: Concurrent[F])
     extends UntypedDestination[F] {
 
   import S3Destination._
 
   def destinationType: DestinationType = DestinationType("s3", 1L)
 
   def sinks: NonEmptyList[ResultSink[F, Unit]] =
-    NonEmptyList.one(csvSink)
-
-  private def csvSink = ResultSink.create[F, Unit, Byte] { (path, _) =>
-    val pipe: Pipe[F, Byte, Unit] =
-      bytes => Stream.eval(for {
-        afile <- ensureAbsFile(path)
-        path = ResourcePath.fromPath(nestResourcePath(afile))
-        key = resourcePathToBlobPath(path)
-        _ <- uploadImpl.upload(bytes, bucket, key)
-      } yield ())
-
-    (RenderConfig.Csv(), pipe)
+    NonEmptyList.of(csvCreateSink, csvAppendSink, csvUpsertSink)
+
+  private def csvCreateSink = ResultSink.create[F, Unit, Byte] { (path, _) =>
+    (RenderConfig.Csv(), consume("CreateSink", path, addTimestamp = false))
+  }
+
+  private def consume(logPrefix: String, path: ResourcePath, addTimestamp: Boolean): Pipe[F, Byte, Unit] = {
+    bytes => Stream.eval(for {
+      afile <- ensureAbsFile(prefixPath, path)
+      postfix <- if (addTimestamp) mkPostfix.map(_.some) else F.pure(none[String])
+      path = ResourcePath.fromPath(nestResourcePath(afile, postfix))
+      key = resourcePathToBlobPath(path)
+      _ <- F.delay(logger.debug(s"[$logPrefix] Uploading ${path.show} to ${key.path.map(_.value).intercalate("/")}"))
+      _ <- uploadImpl.upload(bytes, bucket, key)
+    } yield ())
   }
 
-  private def nestResourcePath(file: AFile): AFile = {
+  private def csvAppendSink = ResultSink.append[F, Unit, Byte](append)
+
+  private def append(appendArgs: ResultSink.AppendSink.Args[Unit])
+      : (RenderConfig[Byte], ∀[Consume[F, AppendEvent[Byte, *], *]]) = {
+    val c = ∀[Consume[F, AppendEvent[Byte, *], *]](consumePipe("AppendSink", appendArgs.path))
+    (RenderConfig.Csv(), c)
+  }
+
+  private def csvUpsertSink = ResultSink.upsert[F, Unit, Byte](upsert)
```

Comment:
Since `Delete` events aren't handled `upsert` behaviour is the same as `append`. 
There's also no tests for `upsert`, maybe remove it?

Reply:
Yes, I agree that it would be better if we could remove it. It's just that when I tested I did get upserts too and I haven't tracked down how to disable this properly. Do you happen to remember?

Reply:
Sorry, I don't quite remember what they do :confused: 
You don't need to disable it, just remove `csvUpsertSink` from `sinks`

Reply:
Hm, but won't it then raise a `Not found` of some sort if the FE thinks it should be there? (I mean let's say if we just enabled increment loading in the FE but not in the BE, then I imagine it will raise a `Not Found` somewhere?)

We could try to remove `Incremental` from here https://github.com/precog/slamx/pull/3754/files#diff-131ea01f35f76af7743997dea3d659d12c073911db1bc6c0972878a0542f5397R235 but I'm not sure if they map 1 to 1 to each other & it will increase risk of running into bugs since it's the first dest that only supports 1 type which means new code paths.

The other reason I included `csvUpsertSink` is that I thought that it does not have any negative consequences. I was thinking that it should always be safe to interpret the contents of the new file as an upsert in any of the 4 combis `upsertSink`/`appendSink` & WriteMode `Replace` / `Append`. I am not sure of this though and I don't have a clear picture of what's the exact difference between them, so please correct me if that assumption is wrong.

(Some example logic near here https://github.com/precog/quasar/blob/10e8c538e255cab29aaa8795d7f9b467308ab76c/impl/src/main/scala/quasar/impl/push/DefaultResultPush.scala#L348 - just putting it here because I vaguely remembered this code but spent some time finding it)

Reply:
The `ResultPush` impl currently couples append sinks to "source driven" pushes, they aren't used for incremental loads from what I can see, so we need the `UpsertSink` for incremental to work at all.

We should document the caveat that it will only be correct when the input is essentially an append log, for the reasons @cryogenian mentioned about not handling deletes.

The `WriteMode` exists to allow resetting the state in the destination. `WriteMode.Replace` means overwrite what is there, start over. `WriteMode.Append` means to add to what exists.

Reply:
@wemrysi Great, thanks - that clears up some things. 

> WriteMode.Replace means overwrite what is there, start over.

What does starting over mean exactly? Would it make sense to encode the fact we are writing in `WriteMode.Replace` mode in the file name somehow, so that the user knows that the contents need to be interpreted differently?

E.g. if it means that we can treat the create events as the "full" load, then we could write the file as `myfile.full.<timestamp>.csv` or something? Or if it means we need to overwrite the last load then we could write the file as `myfile.forgetprevious.<timestamp>.csv`.

> WriteMode.Append means to add to what exists

Is it possible to receive delete events in this case? The types seem to allow it but I don't see a scenario when this would happen.

Reply:
Ok, thanks! I have documented the limitation wrt delete events. Also logging a warning now.

Reply:
If we're not going to actually replace the files, then I think using your suggestion of adding `.full` to the filename sounds good. Maybe when it is `Append`, we add `.increment` or similar to the filename? Then if the user sorts cronologically, they should see `full -> increment -> increment, full -> increment -> increment` etc. What do you think?